src/pyk/proof/proof.py

```diff
@@ -175,9 +176,37 @@ def json(self) -> str:
         return json.dumps(self.dict)
 
     @property
-    def summary(self) -> Iterable[str]:
+    def summary(self) -> ProofSummary:
         subproofs_summaries = [subproof.summary for subproof in self.subproofs]
-        return chain([f'Proof: {self.id}', f'    status: {self.status}'], *subproofs_summaries)
+        return CompositeSummary(ProofSummary(self.id, self.status), subproofs_summaries)
```

Comment:
I agree this is a bit weird. I'd resolve it as follows:

Keep the interface minimal:

```python
class ProofSummary(ABC):

    @property
    @abstractmethod
    def lines(self) -> list[str]:
        ...

    def __str__(self) -> str:
        return '\n'.join(self.lines)
```

Note how, `ProofSummary is truly abstract, there's no way to instantiate it..

Then create a minimal implementation:

```python3
@dataclass
class BaseSummary(ProofSummary):
    id: str
    status: ProofStatus

    @property
    def lines(self) -> list[str]:
        return [f'Proof: {self.id}', f'    status: {self.status}']
```

then

```suggestion
        return CompositeSummary(BaseSummary(self.id, self.status), subproofs_summaries)
```

You can even make `_BaseSummary` private or nest it within `Proof.summary`.